docs/data/base/components/field/field.md

```diff
@@ -0,0 +1,185 @@
+```
+
+```bash pnpm
+pnpm add @base_ui/react
+```
+
+</codeblock>
+
+Once you have the package installed, import the component.
+
+```ts
+import * as Field from '@base_ui/react/Field';
+```
+
+## Anatomy
+
+Fields are implemented using a collection of related components:
+
+- `<Field.Root />` is a top-level component that wraps all other components.
+- `<Field.Control />` renders the control when not using a native Base UI component.
+- `<Field.Label />` renders a label for the control.
+- `<Field.Message />` renders an optional message for the control to describe it or show validation errors.
+- `<Field.Validity />` is an optional render prop component that enables reading raw `ValidityState` to render custom JSX.
+
+```jsx
+<Field.Root>
+  <Field.Control />
+  <Field.Label />
+  <Field.Message />
+  <Field.Validity />
+</Field.Root>
+```
+
+## Accessibility
+
+All Base UI input components are aware of Base UI's `Field` component. The label and description are automatically wired to it when placed inside a `Field.Root`:
+
+```jsx
+<Field.Root>
+  <Checkbox.Root>
+    <Checkbox.Indicator />
+  </Checkbox.Root>
+  <Field.Label>My checkbox</Field.Label>
+  <Field.Message>My description</Field.Message>
+</Field.Root>
+```
+
+When using a native control like `input` or a custom component which is not aware of Base UI's `Field`, use `Field.Control`:
+
+```jsx
+<Field.Root>
+  <Field.Control />
+  <Field.Label>My input</Field.Label>
+  <Field.Message>My description</Field.Message>
+</Field.Root>
+```
+
+The `render` prop allows you to pass a custom component or tag, different from the default of `input`:
+
+```jsx
+<Field.Control render={<select />} />
+```
+
+## Validation
+
+When adding native HTML validation props like `required` or `pattern`, the `show` prop on `Field.Message` shows the message only if the constraint validation fails:
+
+```jsx
+<Field.Root>
+  <Field.Label>My input</Field.Label>
+  <Field.Control required />
+  <Field.Message show="valueMissing" />
+</Field.Root>
+```
+
+The `children` by default is the browser's native message, which is automatically internationalized. You may pass custom `children` instead:
+
+```jsx
+<Field.Root>
+  <Field.Label>My input</Field.Label>
+  <Field.Control required />
+  <Field.Message show="valueMissing">Input is required</Field.Message>
+</Field.Root>
+```
+
+For the list of supported `show` strings, visit [`ValidityState` on MDN](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState#instance_properties).
+
+### Custom validation
+
+In addition to the native HTML constraint validation, you can also add custom validation by passing a function that receives the control's `value` as a first argument on `Field.Root`:
+
+```jsx
+<Field.Root
+  validate={(value) =>
+    value === 'password' ? 'Cannot literally use `password` as your password.' : null
+  }
+>
+  <Field.Control type="password" />
+  <Field.Label>Password</Field.Label>
+  <Field.Message show="customError" />
```

Comment:
Could we allow defining developers define their own error keys?
So `validate` could return a string (or even array of strings) being error keys:
```
 validate={(value) =>
    value === 'password' ? ['literal-password'] : null
  }
```

And `Field.Message` would reference it:
```
<Field.Message show="literal-password" />
```

Reply:
I thought of that, but the lack of auto typesafety for the `show` string was less than ideal. To show multiple custom errors, it currently does require calling an external function twice — once for validation and another for rendering and the `Field.Validity` component:

```js
import * as React from 'react';
import * as Field from '@base_ui/react/Field';

function validate(value: string) {
  const errors: string[] = [];

  if (value.length < 8) {
    errors.push('Password must be at least 8 characters long.');
  }

  if ((value.match(/[A-Z]/g) ?? []).length < 2) {
    errors.push('Password must contain at least 2 uppercase letters.');
  }

  if ((value.match(/[!@#$%^&*]/g) ?? []).length < 2) {
    errors.push(
      'Password must contain at least 2 unique symbols from the set [!@#$%^&*].',
    );
  }

  return errors;
}

function Password() {
  return (
    <Field.Root
      validateOnChange
      validationDebounceMs={0}
      validate={(value) => {
        const password = value as string;
        const errors = validate(password);
        return errors.length ? 'error' : null;
      }}
    >
      <Field.Control type="password" />
      <Field.Label>Password</Field.Label>
      <Field.Message show="customError" render={<ul />}>
        <Field.Validity>
          {(state) => {
            const password = state.value as string;
            const errors = validate(password);

            if (!errors.length) {
              return null;
            }

            return errors.map((error, index) => <li key={index}>{error}</li>);
          }}
        </Field.Validity>
      </Field.Message>
    </Field.Root>
  );
}
```